concordium-consensus/src/Concordium/MultiVersion.hs

```diff
@@ -658,80 +664,82 @@ startupSkov genesis = do
+                                            notifyRegenesis (Just genesisHash)
+                                            mvLog Runner LLTrace "Load configuration done"
+                                            return (Left (newVersion newEConfig, lastFinalizedHeight, nextPV))
+                                        Nothing ->
+                                            case first of
+                                                Nothing -> return (Right Nothing)
+                                                Just newEConfig -> return (Right (Just newEConfig))
+                comp >>= \case
+                    -- We successfully loaded a configuration.
+                    Left (newEConfig@(EVersionedConfiguration newEConfig'), lastFinalizedHeight, nextPV) ->
+                        -- If there is a next protocol then we attempt another loop.
+                        -- If there isn't we attempt to start with the last loaded
+                        -- state as the active state.
+                        case nextPV of
+                            Nothing -> do
+                                mvrLogIO $ activateConfiguration newEConfig
+                                liftSkovUpdate newEConfig' checkForProtocolUpdate
+                            Just nextSPV -> loop nextSPV (Just newEConfig) (vcIndex + 1) (fromIntegral lastFinalizedHeight + 1)
+                    -- We failed to load anything in the first iteration of the
+                    -- loop. Decode the provided genesis and attempt to start the
+                    -- chain.
+                    Right Nothing -> do
+                        logEvent Runner LLTrace "Attempting to decode genesis"
+                        case genesis of
+                            Left genBS -> case runGet getPVGenesisData genBS of
+                                Left err -> do
+                                    logEvent Runner LLError $ "Failed to decode genesis data: " ++ err
+                                    throwM (InvalidGenesisData err)
+                                Right gd -> newGenesis gd 0
                             Right gd -> newGenesis gd 0
-                        Right gd -> newGenesis gd 0
-                -- We loaded some protocol versions. Attempt to start in the
-                -- last one we loaded.
-                Right (Just config@(EVersionedConfiguration newEConfig')) -> do
-                    mvrLogIO $ activateConfiguration config
-                    liftSkovUpdate newEConfig' checkForProtocolUpdate
+                    -- We loaded some protocol versions. Attempt to start in the
+                    -- last one we loaded.
+                    Right (Just config@(EVersionedConfiguration newEConfig')) -> do
+                        mvrLogIO $ activateConfiguration config
+                        liftSkovUpdate newEConfig' checkForProtocolUpdate
+            SConsensusParametersVersion1 -> error "New consensus not implemented yet." -- TODO: implement
```

Comment:
And here.

Reply:
Created this umbrella task for the todos/fixmes in the MultiVersionRuner. https://github.com/Concordium/concordium-node/issues/700